README.md
<h1 align="center">
axum-sessions
</h1>

<p align="center">
🥠 Cookie-based sessions for Axum via async-session.
</p>

<div align="center">
<a href="https://crates.io/crates/axum-sessions">
<img src="https://img.shields.io/crates/v/axum-sessions.svg" />
</a>
<a href="https://docs.rs/axum-sessions">
<img src="https://docs.rs/axum-sessions/badge.svg" />
</a>
<a href="https://github.com/maxcountryman/axum-sessions/actions/workflows/rust.yml">
<img src="https://github.com/maxcountryman/axum-sessions/actions/workflows/rust.yml/badge.svg" />
</a>
<a href='https://coveralls.io/github/maxcountryman/axum-sessions?branch=main'>
<img src='https://coveralls.io/repos/github/maxcountryman/axum-sessions/badge.svg?branch=main' alt='Coverage Status' />
</a>
</div>

## 🎨 Overview

`axum-sessions` is a middleware providing cookie-based sessions for `axum` applications.

- Cryptographically-signed cookies, ensuring integrity and authenticity
- Wraps `async-session`, enabling flexible cookie storage (e.g. `async-sqlx-session`)
- Convenient extractor-based API (i.e. `ReadableSession` and `WritableSession`)
- Can be used as a generic Tower middleware

## 📦 Install

To use the crate in your project, add the following to your `Cargo.toml` file:

```toml
[dependencies]
axum-sessions = "0.5.0"
```

## 🤸 Usage

`axum` applications can use the middleware via the session layer.

### `axum` Example

```rust
use axum::{response::IntoResponse, routing::get, Router};
use axum_sessions::{
    async_session::MemoryStore,
    extractors::{ReadableSession, WritableSession},
    SessionLayer,
};
use rand::Rng;

#[tokio::main]
async fn main() {
    let store = MemoryStore::new();
    let secret = rand::thread_rng().gen::<[u8; 128]>();
    let session_layer = SessionLayer::new(store, &secret).with_secure(false);

    async fn display_handler(session: ReadableSession) -> impl IntoResponse {
        let mut count = 0;
        count = session.get("count").unwrap_or(count);
        format!(
            "Count is: {}; visit /inc to increment and /reset to reset",
            count
        )
    }

    async fn increment_handler(mut session: WritableSession) -> impl IntoResponse {
        let mut count = 1;
        count = session.get("count").map(|n: i32| n + 1).unwrap_or(count);
        session.insert("count", count).unwrap();
        format!("Count is: {}", count)
    }

    async fn reset_handler(mut session: WritableSession) -> impl IntoResponse {
        session.destroy();
        "Count reset"
    }

    let app = Router::new()
        .route("/", get(display_handler))
        .route("/inc", get(increment_handler))
        .route("/reset", get(reset_handler))
        .layer(session_layer);

    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

You can find this [example][counter-example] as well as other example projects in the [example directory][examples].

## Session authentication via `axum-login`

For user session management and authentication and authorization specifically please see [`axum-login`](https://github.com/maxcountryman/axum-login).

See the [crate documentation][docs] for more usage information.

[counter-example]: https://github.com/maxcountryman/axum-sessions/tree/main/examples/counter
[examples]: https://github.com/maxcountryman/axum-sessions/tree/main/examples
[docs]: https://docs.rs/axum-sessions


CHANGELOG.md
# 0.5.0

**BREAKING CHANGES**:

- Resist session name fingerprinting [PR #36](https://github.com/maxcountryman/axum-sessions/pull/36)

**OTHER CHANGES**:

- Allow setting HttpOnly of cookie [PR #30](https://github.com/maxcountryman/axum-sessions/pull/30)

# 0.4.1

- Update axum to v0.6.0

# 0.4.0

- Avoid storing cookie when not required [PR #15](https://github.com/maxcountryman/axum-sessions/pull/15)

# 0.3.2

- Search every Cookie header for session cookie [PR #14](https://github.com/maxcountryman/axum-sessions/pull/14)

# 0.3.1

- Derive `Debug` for `WritableSession` and `ReadableSession` ensuring consistency with `Session`

# 0.3.0

- Session regeneration support [PR #6](https://github.com/maxcountryman/axum-sessions/pull/6)

# 0.2.0

- On session destroy, unset cookie on client [PR #4](https://github.com/maxcountryman/axum-sessions/pull/4)

# 0.1.1

- Handle multiple cookie values [PR #2](https://github.com/maxcountryman/axum-sessions/pull/2)

# 0.1.0

- Initial release :tada:


src/lib.rs
//! axum-sessions is a middleware providing cookie-based sessions for axum
//! applications.
//!
//! [`SessionLayer`] provides client sessions via [`async_session`]. Sessions
//! are backed by cryptographically signed cookies. These cookies are generated
//! when they're not found or are otherwise invalid. When a valid, known cookie
//! is received in a request, the session is hydrated from this cookie. The
//! middleware provides sessions via [`SessionHandle`]. Handlers use the
//! [`ReadableSession`](crate::extractors::ReadableSession) and
//! [`WritableSession`](crate::extractors::WritableSession) extractors to read
//! from and write to sessions respectively.
//!
//! # Example
//!
//! Using the middleware with axum is straightforward:
//!
//! ```rust,no_run
//! use axum::{routing::get, Router};
//! use axum_sessions::{
//!     async_session::MemoryStore, extractors::WritableSession, PersistencePolicy, SessionLayer,
//! };
//!
//! #[tokio::main]
//! async fn main() {
//!     let store = async_session::MemoryStore::new();
//!     let secret = b"..."; // MUST be at least 64 bytes!
//!     let session_layer = SessionLayer::new(store, secret);
//!
//!     async fn handler(mut session: WritableSession) {
//!         session
//!             .insert("foo", 42)
//!             .expect("Could not store the answer.");
//!     }
//!
//!     let app = Router::new().route("/", get(handler)).layer(session_layer);
//!
//!     axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
//!         .serve(app.into_make_service())
//!         .await
//!         .unwrap();
//! }
//! ```
//!
//! This middleware may also be used as a generic Tower middleware by making use
//! of the [`SessionHandle`] extension:
//!
//! ```rust
//! use std::convert::Infallible;
//!
//! use axum::http::header::SET_COOKIE;
//! use axum_sessions::{SessionHandle, SessionLayer};
//! use http::{Request, Response};
//! use hyper::Body;
//! use rand::Rng;
//! use tower::{Service, ServiceBuilder, ServiceExt};
//!
//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Infallible> {
//!     let session_handle = request.extensions().get::<SessionHandle>().unwrap();
//!     let session = session_handle.read().await;
//!     // Use the session as you'd like.
//!
//!     Ok(Response::new(Body::empty()))
//! }
//!
//! # #[tokio::main]
//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
//! let store = async_session::MemoryStore::new();
//! let secret = rand::thread_rng().gen::<[u8; 128]>();
//! let session_layer = SessionLayer::new(store, &secret);
//!
//! let mut service = ServiceBuilder::new()
//!     .layer(session_layer)
//!     .service_fn(handle);
//!
//! let request = Request::builder().body(Body::empty()).unwrap();
//!
//! let response = service.ready().await?.call(request).await?;
//!
//! assert_eq!(
//!     response
//!         .headers()
//!         .get(SET_COOKIE)
//!         .unwrap()
//!         .to_str()
//!         .unwrap()
//!         .split("=")
//!         .collect::<Vec<_>>()[0],
//!     "sid"
//! );
//!
//! # Ok(())
//! # }
//! ```

#![deny(missing_docs)]

pub mod extractors;
mod session;

pub use async_session;
pub use axum_extra::extract::cookie::SameSite;

pub use self::session::{PersistencePolicy, Session, SessionHandle, SessionLayer};


src/extractors.rs
//! Extractors for sessions.

use std::ops::{Deref, DerefMut};

use axum::{async_trait, extract::FromRequestParts, http::request::Parts, Extension};
use tokio::sync::{OwnedRwLockReadGuard, OwnedRwLockWriteGuard};

use crate::SessionHandle;

/// An extractor which provides a readable session. Sessions may have many
/// readers.
#[derive(Debug)]
pub struct ReadableSession {
    session: OwnedRwLockReadGuard<async_session::Session>,
}

impl Deref for ReadableSession {
    type Target = OwnedRwLockReadGuard<async_session::Session>;

    fn deref(&self) -> &Self::Target {
        &self.session
    }
}

#[async_trait]
impl<S> FromRequestParts<S> for ReadableSession
where
    S: Send + Sync,
{
    type Rejection = std::convert::Infallible;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let Extension(session_handle): Extension<SessionHandle> =
            Extension::from_request_parts(parts, state)
                .await
                .expect("Session extension missing. Is the session layer installed?");
        let session = session_handle.read_owned().await;

        Ok(Self { session })
    }
}

/// An extractor which provides a writable session. Sessions may have only one
/// writer.
#[derive(Debug)]
pub struct WritableSession {
    session: OwnedRwLockWriteGuard<async_session::Session>,
}

impl Deref for WritableSession {
    type Target = OwnedRwLockWriteGuard<async_session::Session>;

    fn deref(&self) -> &Self::Target {
        &self.session
    }
}

impl DerefMut for WritableSession {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.session
    }
}

#[async_trait]
impl<S> FromRequestParts<S> for WritableSession
where
    S: Send + Sync,
{
    type Rejection = std::convert::Infallible;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let Extension(session_handle): Extension<SessionHandle> =
            Extension::from_request_parts(parts, state)
                .await
                .expect("Session extension missing. Is the session layer installed?");
        let session = session_handle.write_owned().await;

        Ok(Self { session })
    }
}


examples/counter/src/main.rs
//! Run with
//!
//! ```not_rust
//! cd examples && cargo run -p example-counter
//! ```

use axum::{response::IntoResponse, routing::get, Router};
use axum_sessions::{
    async_session::MemoryStore,
    extractors::{ReadableSession, WritableSession},
    SessionLayer,
};
use rand::Rng;

#[tokio::main]
async fn main() {
    let store = MemoryStore::new();
    let secret = rand::thread_rng().gen::<[u8; 128]>();
    let session_layer = SessionLayer::new(store, &secret).with_secure(false);

    async fn display_handler(session: ReadableSession) -> impl IntoResponse {
        let mut count = 0;
        count = session.get("count").unwrap_or(count);
        format!(
            "Count is: {}; visit /inc to increment and /reset to reset",
            count
        )
    }

    async fn increment_handler(mut session: WritableSession) -> impl IntoResponse {
        let mut count = 1;
        count = session.get("count").map(|n: i32| n + 1).unwrap_or(count);
        session.insert("count", count).unwrap();
        format!("Count is: {}", count)
    }

    async fn reset_handler(mut session: WritableSession) -> impl IntoResponse {
        session.destroy();
        "Count reset"
    }

    let app = Router::new()
        .route("/", get(display_handler))
        .route("/inc", get(increment_handler))
        .route("/reset", get(reset_handler))
        .layer(session_layer);

    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
